# 01. Importing libraries

In [7]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Importing data

In [8]:
# Import Orders data from Prepared Data folder
df_ords = pd.read_csv(r'C:\Users\lrutl\OneDrive\Desktop\Python Fundamentals for Data Analysts\Instacart Basket Analysis\02 Data\Prepared Data\orders_wrangled.csv', index_col = False)

In [9]:
path = r'C:\Users\lrutl\OneDrive\Desktop\Python Fundamentals for Data Analysts\Instacart Basket Analysis'

In [10]:
df_ords.head()

Unnamed: 0  order_id  customer_id  order_number  orders_day_of_week  \
0           0   2539329            1             1                   2   
1           1   2398795            1             2                   3   
2           2    473747            1             3                   3   
3           3   2254736            1             4                   4   
4           4    431534            1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [11]:
# Import Products data from Original Data folder
df_prods = pd.read_csv(r'C:\Users\lrutl\OneDrive\Desktop\Python Fundamentals for Data Analysts\Instacart Basket Analysis\02 Data\Original Data\products.csv', index_col = False)

In [12]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

# 03. Mixed-type data

In [13]:
# Create test dataframe
df_test = pd.DataFrame()

In [14]:
# Create a mixed type column
df_test['mix'] = ['a','b',1,True]

In [15]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [16]:
# Check for mixed type data columns
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [17]:
# Convert mixed type data column to string
df_test['mix'] = df_test['mix'].astype('str')

# 04. Missing values

In [18]:
# Find missing values in df_prods
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [19]:
# Create new dataframe with only null values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [20]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [21]:
df_prods.shape

(49693, 5)

In [22]:
# Create new dataframe to remove null rows
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [23]:
df_prods_clean.shape

(49677, 5)

# 05. Duplicates

In [26]:
# Find full duplicates in df_prods_clean - create subset of only duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [27]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [28]:
# Addressing duplicates
df_prods_clean.shape

(49677, 5)

In [29]:
# Create new dataframe that does NOT include identified duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [30]:
df_prods_clean_no_dups.shape

(49672, 5)

# 06. Tidying up and exporting changes

In [31]:
# Export df_prods_clean_no_dups dataframe as products_checked.csv to prepared data folder
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Exercise 4.5

In [32]:
# 2.a. Run df.describe() on df_prods dataframe
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

2.b. The max price is very high compared to the average. This means there are likely outliers in the prices column.

In [33]:
# 3. Check for mixed type data columns in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

4. No mixed type data based on the above check

In [34]:
# 5. Check for missing values in df_ords
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
customer_id                    0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [35]:
df_ords.shape

(3421083, 7)

In [36]:
df_ords.describe()

Unnamed: 0      order_id   customer_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

5.b. 206,209 missing values in days_since_prior_order column. First, I thought this might mean there were no days since the prior order but after looking at the df_ords.describe() output, those are accounted for as "0" rather than NaN. This makes me think that for some reason, the data was not recorded in the 206,209 instances of missing values. I think the best way to address this is to delete the rows with missing data because it's only 6% of the total rows in the dataframe.

In [37]:
# 6. Create new dataframe to remove null rows
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [39]:
df_ords_clean.shape

(3214874, 7)

6.b. I created a new dataframe that excludes rows with null values because I do not want to delete data from the original file, and excluding it for my analysis will have a minimal impact as it is only 6% of the total data in this dataframe.

In [40]:
# 7. Find duplicate values in df_ords_clean
df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [41]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, customer_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

7.b. No full duplicates in df_ords_clean.

8. No action needed because there are no full duplicates.

In [42]:
# 9. Export df_prods_clean_no_dups dataframe as products_checked.csv to prepared data folder
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [43]:
# 9.b. Export df_ords_clean dataframe as orders_checked.csv to prepared data folder
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))